htdocs/lib/js/ace/mode-r.js

```diff
@@ -116,7 +110,92 @@ define("ace/mode/r", function(require, exports, module)
+"use strict";
+
+
+var RCompletions = function() {
+  
+};
+
+(function() {
+    this.getCompletions = function(state, session, pos, prefix, callback) {
+          var that = this;
+          rcloud.get_completions('R', session.getValue(), session.getDocument().positionToIndex(pos))
+                                 .then(function(ret) { 
+                                   ret.forEach(function(x) { x.completer = that; }); 
+                                  return ret;
+                                 })
+                                .then(function(ret) {
+                                    callback(null, ret);
+                                  });
+      };
+      
+   this.identifierRegexps = [/[a-zA-Z_0-9:\/\$\-\.\u00A2-\uFFFF]/];
+      
+   this.insertMatch = function(editor, data) {
+      var completions = editor.completer.completions;
+      var session = editor.getSession();
+      var pos = editor.getCursorPosition();
+      var line = session.getLine(pos.row);
+      var left = line.substr(0,pos.column);
+      var right = line.substr(pos.column, line.length);
+      
+      var removeToLeft = function(editor, howMany) {
+          var range = editor.selection.getRange();
+          range.start.column -= howMany;
+          editor.session.remove(range);
+      };
+      
+      if ( completions ) {
+        // Note: filterText may contain initial text and any extra characters that user typed in to filter
+        // the set of available completions produced when autocomplete dialog was created.
+        if(completions.filterText) {
+          if( left.endsWith(completions.filterText) && data.value.startsWith(completions.filterText) ) {
+            // Avoid unnecessary autocompletion
+            var replacementTail = data.value.substr(completions.filterText.length, data.value.length);
+            if(right.startsWith(replacementTail)) {
+              var gotoRange = editor.selection.getRange();
+              gotoRange.end.column += replacementTail.length;
+              gotoRange.start.column += replacementTail.length;
+              editor.selection.setSelectionRange(gotoRange);
+              return;
+            }
+          }
+          removeToLeft(editor, completions.filterText.length);
+          left = left.substr(0, left.length - completions.filterText.length);
+        }
+        if( left.length > 0 ) {
+          // Handle a case when the selected completion starts ahead current range/position in the editor
+          var lookbackIndex = 0;
+          var prefix = left.substr(lookbackIndex, left.length);
+          while(lookbackIndex < left.length && !data.value.startsWith(prefix)) {
+            lookbackIndex += 1;
+            prefix = left.substr(left.length-lookbackIndex, left.length);
+          }
+          if(left.length > lookbackIndex) {
+            removeToLeft(editor, left.length - lookbackIndex);
+          }
+        }
+      }
```

Comment:
This is the code which seems excessively complicated to me. 

Would you need to do any of this if you had the positions from R's autocomplete? It seems to me that if you just replaced the region defined by those positions, with the selected text, and put the cursor at the end, then there would be no need to search for anything or guess what was "meant" by a substitution.